### DISCUSSION TITLE
Input should be a valid datetime, second fraction value is more than 6 digits long

### DISCUSSION THREAD

**Discussion by hf-kklein** [pydantic/pydantic#6411] [Question]
When binding a server response to my Pydantic (v2) model, I get the following error:

> Input should be a valid datetime, second fraction value is more than 6 digits long [type=datetime_parsing, input_value='9999-12-31T23:59:59.9999999+00:00', input_type=str]

The `input_value` is a `DateTimeOffset.MaxValue` from a C#/ASP.NET server; I cannot change it.

What's the easiest way to drop the 7th `9`? Feels kind of wrong to parse the datetime using a regex first 😀 Is there a builtin way?

The [documentation](https://docs.pydantic.dev/dev-v2/usage/validation_errors/#datetime_parsing) does not explain how to ignore the too many decimal places.

### Workaround:
Using regex to "parse" the datetime first:
```python
_datetime_pattern = re.compile(
    r"^(?P<dateandtime>.*)(?P<upto6decimals>\.\d{0,6})(?P<submicroseconds>\d*)(?P<therest>.*$)"
)
"""
a pattern that matches the .9999999 in the submicroseconds group
"""
class MyModel(BaseModel):
    my_datetime: datetime 
   
    @field_validator("my_datetime", mode="before")
    def drop_too_many_second_fractions(cls, v):
        """
        prevents 'Input should be a valid datetime, second fraction value is more than 6 digits long'
        # https://github.com/pydantic/pydantic/discussions/6411
        """
        if match := _datetime_pattern.match(v):
            return match.group("dateandtime") + match.group("upto6decimals") + match.group("therest")
        return v
```

**Comment by kevin-paulson-mindbridge-ai** [comment]
As an alternative you could use `python-dateutil`, so something like:
```python
from datetime import datetime
from dateutil.parser import isoparse
from pydantic import field_validator

# more code here

    @field_validator("creation_date", mode="before")
    def _convert_long_datetime(cls, v: str) -> datetime:
        """Converts an iso8601 datetime string to datetime to avoid this more than 6 digits issue:
        creationDate
        Input should be a valid datetime, second fraction value is more than 6 digits long [type=datetime_parsing, input_value='2023-07-07T15:08:44.966381545Z', input_type=str]
          For further information visit https://errors.pydantic.dev/2.1.2/v/datetime_parsing
        """
        return isoparse(v)
```

I think pydantic should round or truncate the second fractional digits to 6 digits. My understanding is although python limits to 6 digits of second fractional digits, the standard (ISO 8601) doesn't have that limitation and other tools can present more than 6 (at least the private api I'm working with today does this 🙂).

**Comment by adriangb** [comment]
The issue I see is that truncating is a loss of information. At the very least, it should be opt-in. But I do think we should provide some way to handle this that doesn't involve a regex. @samuelcolvin what do you think?

**Reply by ybressler** [reply]
An ideal solution would look something like this:
```py3
from pydantic.magic import parse_datetime


def wrap_magic_parse_datetime(...):
    """
    Calls a magic function which parses datetime.

    Note: We are going to pass in an argument which truncates our timestamp (cause we wanna).
    """
    return parse_datetime(..., truncate=True)


class MyModel(BaseModel):
    my_datetime: Annotated[datetime, BeforeValidator(parse_datetime)]
```

**Comment by samuelcolvin** [comment]
See https://github.com/chronotope/chrono/issues/171#issuecomment-313695814, I think I was wrong.

It seems both iso8601 and rfc3339 allow arbitrary precision in second fractions.

I agree we should allow a way to parse arbitrary precision.

The big question is, what should we make the default?

**Reply by samuelcolvin** [reply]
https://github.com/pydantic/speedate/issues/40

**Comment by samuelcolvin** [comment]
Loving discussion to the issue #6512.
